src/utils.py

```diff
@@ -0,0 +1,135 @@
+import sys
+import imp
+import json
+from os.path import join, exists
+from collections import OrderedDict
+
+ENCODING = 'ISO-8859-1'
+
+
+def is_empty(val):
+    """Check if a variable is an empty string or an empty list."""
+    return val == "" or val == []
+
+
+def clean_input(prompt="", split_char='', ignore_empty=False, dtype=None):
+    """Clean the user-input from the CLI before adding it."""
+    while True:
+        val = input(prompt).strip()
+        # split to list type if split_char specified
+        if split_char != "":
+            val = [v.strip() for v in val.split(split_char) if v.strip() != ""]
+        # do not ignore empty input if not allowed
+        if not ignore_empty and is_empty(val):
+            print("\tError: empty input. Need one or more values.\n")
+            continue
+        # ensure correct input datatype if specified
+        if not is_empty(val) and dtype is not None:
+            try:
+                if not type(eval(val)) == dtype:
+                    print("\tError: input doesn't match required type ", dtype, "\n")
+                    continue
+            except:
+                print("\tError: illegal argument. Input type should be ", dtype, "\n")
+                continue
+        break
+    return val
+
+
+def open_fr(file_name, encoding=ENCODING, encode=True):
+    """Open file for reading respecting Python version and OS differences.
+
+    Sets newline to Linux line endings on Windows and Python 3
+    When encode=False does not set encoding on nix and Python 3 to keep as bytes
+    """
+    if sys.version_info >= (3, 0, 0):
+        if os.name == 'nt':
+            file_obj = io.open(file_name, 'r', newline='', encoding=encoding)
+        else:
+            if encode:
+                file_obj = io.open(file_name, "r", encoding=encoding)
+            else:
+                file_obj = io.open(file_name, "r")
+    else:
+        file_obj = io.open(file_name, "r", encoding=encoding)
+    return file_obj
+
+
+def read_json(json_file):
+    """Read Json dataset package files
+
+    Load each json and get the appropriate encoding for the dataset
+    Reload the json using the encoding to ensure correct character sets
+    """
+    json_object = OrderedDict()
+    json_file_encoding = None
+
+    try:
+        file_obj = open_fr(json_file)
+        json_object = json.load(file_obj)
+        if "encoding" in json_object:
+            json_file_encoding = json_object['encoding']
+        file_obj.close()
+    except ValueError:
+        pass
+
+    # Reload json using encoding if available
+    try:
+        if json_file_encoding:
+            file_obj = open_fr(json_file, encoding=json_file_encoding)
+        else:
+            file_obj = open_fr(json_file)
+        json_object = json.load(file_obj)
+        file_obj.close()
+    except ValueError:
+        pass
+
+    if type(json_object) is dict and "version" in json_object.keys():
+        return json_object["version"]
+    return None
```

Comment:
Code/function looks good shipped from [read_json](https://github.com/weecology/retriever/blob/master/retriever/lib/load_json.py#L16) in retriever/master
lines removed `json_file = str(json_file) + ".json"` and looks like we are removing some unwanted [code]( [read_json](https://github.com/weecology/retriever/blob/master/retriever/lib/load_json.py#L46)) 
I think I will need some clarification on this

Reply:
@henrykironde I didn't quite understand. Could you please elaborate?

Reply:
I am making some notes on the code for everyone. For example, if a function was moved from the retriever directory and nothing was changed, I want to mark it as reviewed. The above is a function that was moved from  [read_json retriever](https://github.com/weecology/retriever/blob/master/retriever/lib/load_json.py#L16), however it leaves out some code. I am trying to understand that part and we can talk more about this during your meeting when you call